File: dream-fronend/src/app/thongke/thongke.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface TopSanPhamResponse {
  tenSanPham: string;  // Tên sản phẩm
  tongSoLuong: number; // Tổng số lượng bán được
}

export interface ThongKeHomNayResponse {
  soHoaDon: number;         // Số hóa đơn trong ngày hôm nay
  tongDoanhThu: number;     // Tổng doanh thu trong ngày hôm nay
  soKhachHang: number;      // Số khách hàng trong ngày hôm nay
}
export interface ThongKeResponse {
  soHoaDon: number;
  tongDoanhThu: number;
  soKhachHang: number;
}

export interface ThongKeThangResponse {
  thang: number; // Có thể là tháng hoặc năm, tùy vào API trả về
  tongDoanhThu: number;
}
export interface ThongKeThangNayResponse {
  ngay: number; // Có thể là tháng hoặc năm, tùy vào API trả về
  tongDoanhThu: number;
}
export interface ThongKeRequest {
  startDate: string | null;
  endDate: string | null;
}

@Injectable({
  providedIn: 'root',
})
export class ThongKeService {
  private apiUrl = 'http://localhost:8080/api/thong-ke';

  constructor(private http: HttpClient) {}

  // Lấy dữ liệu thống kê tổng quan
  thongKeTongQuan(type: string): Observable<ThongKeResponse> {
    return this.http.get<ThongKeResponse>(`${this.apiUrl}/${type}`);
  }

  // Lấy dữ liệu thống kê doanh thu từng tháng trong năm nay
  thongKeTungThangNam(): Observable<ThongKeThangResponse[]> {
    return this.http.get<ThongKeThangResponse[]>(`${this.apiUrl}/nam-nay/thang`);
  }

  // Lấy dữ liệu thống kê doanh thu theo từng năm
  thongKeTungNam(): Observable<ThongKeThangResponse[]> {
    return this.http.get<ThongKeThangResponse[]>(`${this.apiUrl}/tat-ca/nam`);
  }

  // Lấy dữ liệu thống kê doanh thu theo từng ngày trong tháng
  thongKeTungNgayTrongThang(): Observable<ThongKeThangNayResponse[]> {
    return this.http.get<ThongKeThangNayResponse[]>(`${this.apiUrl}/thang-nay/ngay`);
  }
  // Lấy dữ liệu thống kê doanh thu ngày hôm nay
  thongKeHomNay(): Observable<ThongKeHomNayResponse> {
    return this.http.get<ThongKeHomNayResponse>(`${this.apiUrl}/hom-nay`);
  }


   // Lấy top sản phẩm bán chạy trong ngày hôm nay
   topSanPhamHomNay(page: number, size: number): Observable<TopSanPhamResponse[]> {
    return this.http.get<TopSanPhamResponse[]>(`${this.apiUrl}/hom-nay/top-san-pham?page=${page}&size=${size}`);
  }

  // Lấy top sản phẩm bán chạy trong tháng này
  topSanPhamThangNay(page: number, size: number): Observable<TopSanPhamResponse[]> {
    return this.http.get<TopSanPhamResponse[]>(`${this.apiUrl}/thang-nay/top-san-pham?page=${page}&size=${size}`);
  }

  // Lấy top sản phẩm bán chạy trong năm nay
  topSanPhamNamNay(page: number, size: number): Observable<TopSanPhamResponse[]> {
    return this.http.get<TopSanPhamResponse[]>(`${this.apiUrl}/nam-nay/top-san-pham?page=${page}&size=${size}`);
  }

  // Lấy top sản phẩm bán chạy tất cả thời gian
  topSanPhamTatCa(page: number, size: number): Observable<TopSanPhamResponse[]> {
    return this.http.get<TopSanPhamResponse[]>(`${this.apiUrl}/tat-ca/top-san-pham?page=${page}&size=${size}`);
  }
}

File: dream-backend/src/main/java/com/example/dreambackend/services/thongke/ThongKeService.java
package com.example.dreambackend.services.thongke;

import com.example.dreambackend.repositories.HoaDonChiTietRepository;
import com.example.dreambackend.repositories.HoaDonRepository;
import com.example.dreambackend.responses.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.YearMonth;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class ThongKeService {

    @Autowired
    private HoaDonRepository hoaDonRepository;
    @Autowired
    private HoaDonChiTietRepository hoaDonChiTietRepository;

    public ThongKeResponse thongKeTongQuan(String type) {
        LocalDate startDate = null;
        LocalDate endDate = null;

        switch (type) {
            case "hom-nay":
                startDate = LocalDate.now();
                endDate = LocalDate.now();
                break;
            case "thang-nay":
                YearMonth currentMonth = YearMonth.now();
                startDate = currentMonth.atDay(1);  // Ngày đầu tháng
                endDate = currentMonth.atEndOfMonth();  // Ngày cuối tháng
                break;
            case "nam-nay":
                int currentYear = LocalDate.now().getYear();
                startDate = LocalDate.of(currentYear, 1, 1);
                endDate = LocalDate.of(currentYear, 12, 31);
                break;
            case "tat-ca":
                // Không thiết lập startDate và endDate
                break;
            default:
                throw new IllegalArgumentException("Invalid type: " + type);
        }

        return hoaDonRepository.getTongQuan(startDate, endDate);
    }

    public List<ThongKeThangResponse> thongKeTungThang() {
        return hoaDonRepository.getDoanhThuTungThang();
    }
    public List<ThongKeThangResponse> thongKeTungNam() {
        List<Object[]> results = hoaDonRepository.getDoanhThuTungNam();
        return results.stream()
                .map(obj -> new ThongKeThangResponse((Integer) obj[0], (Double) obj[1]))
                .collect(Collectors.toList());
    }
    // Thống kê doanh thu từng ngày trong tháng
    public List<ThongKeThangNayResponse> thongKeTungNgayTrongThang() {
        return hoaDonRepository.getDoanhThuTungNgayTrongThang();
    }
    // Thống kê doanh thu ngày hôm nay
    public ThongKeHomNayResponse thongKeHomNay() {
        return hoaDonRepository.getDoanhThuHomNay();
    }
    // Thống kê sản phẩm bán chạy nhất trong ngày hôm nay
    public List<TopSanPhamResponse> topSanPhamHomNay() {
        return hoaDonChiTietRepository.getTopSanPhamHomNay(PageRequest.of(0, 10)).getContent();
    }

    // Thống kê sản phẩm bán chạy nhất trong tháng này
    public List<TopSanPhamResponse> topSanPhamThangNay() {
        LocalDate startDate = LocalDate.now().withDayOfMonth(1); // Ngày đầu tháng
        LocalDate endDate = LocalDate.now().withDayOfMonth(LocalDate.now().lengthOfMonth()); // Ngày cuối tháng
        return hoaDonChiTietRepository.getTopSanPhamThangNay(PageRequest.of(0, 10), startDate, endDate).getContent();
    }

    // Thống kê sản phẩm bán chạy nhất trong năm nay
    public List<TopSanPhamResponse> topSanPhamNamNay() {
        return hoaDonChiTietRepository.getTopSanPhamNamNay(PageRequest.of(0, 10)).getContent();
    }

    // Thống kê sản phẩm bán chạy nhất tất cả thời gian
    public List<TopSanPhamResponse> topSanPhamTatCa() {
        return hoaDonChiTietRepository.getTopSanPhamTatCa(PageRequest.of(0, 10)).getContent();
    }
}
